File: golang/internal/agent/risk_identification.go
package agent

import (
	"encoding/json"
	"fmt"
	"log"

	"github.com/bcefghj/multi-agent-contract-review/golang/internal/config"
	"github.com/bcefghj/multi-agent-contract-review/golang/internal/model"
	"github.com/bcefghj/multi-agent-contract-review/golang/internal/rule"
)

type RiskIdentificationAgent struct {
	llmClient  *config.LLMClient
	ruleEngine *rule.RuleEngine
}

func NewRiskIdentificationAgent(client *config.LLMClient, engine *rule.RuleEngine) *RiskIdentificationAgent {
	return &RiskIdentificationAgent{llmClient: client, ruleEngine: engine}
}

func (a *RiskIdentificationAgent) Name() string { return "RiskIdentificationAgent" }

func (a *RiskIdentificationAgent) Process(state *model.ReviewState) error {
	log.Printf("[%s] 开始处理 reviewId=%s", a.Name(), state.ReviewID)

	if len(state.Clauses) == 0 {
		state.OverallRiskLevel = model.RiskNone
		state.RiskSummary = "无条款可供分析"
		return nil
	}

	ruleFindings := a.ruleEngine.ScanClauses(state.Clauses)

	var llmFindings []model.RiskFinding
	llmResult, err := a.analyzeWithLLM(state.Clauses)
	if err != nil {
		log.Printf("[%s] LLM分析失败: %v", a.Name(), err)
		state.Errors = append(state.Errors, fmt.Sprintf("风险LLM分析失败: %v", err))
	} else {
		llmFindings = llmResult
	}

	merged := mergeFindings(ruleFindings, llmFindings)
	overall := calculateOverallRisk(merged)

	state.RiskFindings = merged
	state.OverallRiskLevel = overall
	state.RiskSummary = generateRiskSummary(merged, overall)
	state.NeedsHumanReview = overall == model.RiskHigh

	log.Printf("[%s] 完成：%d个发现，整体=%s", a.Name(), len(merged), overall)
	return nil
}

func (a *RiskIdentificationAgent) analyzeWithLLM(clauses []model.Clause) ([]model.RiskFinding, error) {
	text := ""
	for _, c := range clauses {
		text += fmt.Sprintf("[条款ID: %s] %s\n%s\n\n", c.ID, c.Title, c.Content)
	}

	systemPrompt := `你是资深法律风险评估专家。输出JSON：
{"findings":[{"clause_id":"","risk_level":"high/medium/low/none","risk_type":"","description":"","rationale":""}],"overall_risk_level":"","risk_summary":""}`

	response, err := a.llmClient.Chat(systemPrompt, "分析条款风险：\n\n"+text)
	if err != nil {
		return nil, err
	}

	jsonStr := extractJSON(response)
	var result struct {
		Findings []struct {
			ClauseID    string `json:"clause_id"`
			RiskLevel   string `json:"risk_level"`
			RiskType    string `json:"risk_type"`
			Description string `json:"description"`
			Rationale   string `json:"rationale"`
		} `json:"findings"`
	}

	if err := json.Unmarshal([]byte(jsonStr), &result); err != nil {
		return nil, fmt.Errorf("JSON解析失败: %w", err)
	}

	var findings []model.RiskFinding
	for _, f := range result.Findings {
		findings = append(findings, model.RiskFinding{
			ClauseID:    f.ClauseID,
			RiskLevel:   model.RiskLevel(f.RiskLevel),
			RiskType:    f.RiskType,
			Description: f.Description,
			Rationale:   f.Rationale,
		})
	}
	return findings, nil
}

func mergeFindings(rule, llm []model.RiskFinding) []model.RiskFinding {
	seen := make(map[string]model.RiskFinding)
	riskPriority := map[model.RiskLevel]int{
		model.RiskHigh: 3, model.RiskMedium: 2, model.RiskLow: 1, model.RiskNone: 0,
	}

	for _, f := range rule {
		key := f.ClauseID + "|" + f.RiskType
		seen[key] = f
	}
	for _, f := range llm {
		key := f.ClauseID + "|" + f.RiskType
		if existing, ok := seen[key]; ok {
			if riskPriority[f.RiskLevel] > riskPriority[existing.RiskLevel] {
				seen[key] = f
			}
		} else {
			seen[key] = f
		}
	}

	result := make([]model.RiskFinding, 0, len(seen))
	for _, f := range seen {
		result = append(result, f)
	}
	return result
}

func calculateOverallRisk(findings []model.RiskFinding) model.RiskLevel {
	for _, f := range findings {
		if f.RiskLevel == model.RiskHigh {
			return model.RiskHigh
		}
	}
	for _, f := range findings {
		if f.RiskLevel == model.RiskMedium {
			return model.RiskMedium
		}
	}
	for _, f := range findings {
		if f.RiskLevel == model.RiskLow {
			return model.RiskLow
		}
	}
	return model.RiskNone
}

func generateRiskSummary(findings []model.RiskFinding, overall model.RiskLevel) string {
	high, medium, low := 0, 0, 0
	for _, f := range findings {
		switch f.RiskLevel {
		case model.RiskHigh:
			high++
		case model.RiskMedium:
			medium++
		case model.RiskLow:
			low++
		}
	}
	return fmt.Sprintf("整体风险：%s。%d个高风险、%d个中风险、%d个低风险。", overall, high, medium, low)
}

File: python/src/agents/risk_identification.py
"""
风险识别Agent — 多Agent流水线第二步。

职责：
1. 对每个条款进行风险评分（高/中/低）
2. 识别不合理条款（无限责任、单方解除权等）
3. 买卖方偏向分析
4. 综合风险评估

采用规则引擎 + 大模型双引擎架构：
- 规则引擎：处理确定性风险检查（速度快、100%确定）
- 大模型：处理需要语义理解的风险分析
"""

from __future__ import annotations

import json
import logging

from langchain_core.messages import HumanMessage, SystemMessage

from ..config import get_llm
from ..pipeline.state import (
    Clause,
    ContractReviewState,
    ReviewStatus,
    RiskFinding,
    RiskLevel,
)

logger = logging.getLogger(__name__)

SYSTEM_PROMPT = """你是一名资深的法律风险评估专家。你的任务是分析合同条款中的潜在风险。

请对每个条款进行风险评估，输出严格的JSON格式（不要输出其他内容）：

{
  "findings": [
    {
      "clause_id": "条款ID",
      "risk_level": "high/medium/low/none",
      "risk_type": "风险类型",
      "description": "风险描述",
      "buyer_impact": "对买方/甲方的影响",
      "seller_impact": "对卖方/乙方的影响",
      "rationale": "判断依据"
    }
  ],
  "overall_risk_level": "high/medium/low/none",
  "risk_summary": "风险总结（50字以内）"
}

常见风险类型包括：
- unlimited_liability: 无限责任
- unilateral_termination: 单方解除权
- unfair_penalty: 不合理违约金
- vague_terms: 模糊条款
- missing_protection: 缺少保护条款
- ip_risk: 知识产权风险
- payment_risk: 付款风险
- confidentiality_risk: 保密风险
- jurisdiction_risk: 管辖权风险
- auto_renewal_risk: 自动续约风险

风险等级判断标准：
- high: 可能导致重大经济损失或法律纠纷
- medium: 存在潜在风险，建议修改
- low: 轻微风险，建议关注
- none: 无明显风险"""


# ── 规则引擎预筛（第一层：确定性检查）──────────────────────────

RISK_RULES = [
    {
        "name": "unlimited_liability",
        "keywords": ["无限责任", "不限于", "承担全部", "无上限", "不设限"],
        "risk_level": RiskLevel.HIGH,
        "description": "条款包含无限责任相关表述",
    },
    {
        "name": "unilateral_termination",
        "keywords": ["单方解除", "单方终止", "有权随时", "任意解除"],
        "risk_level": RiskLevel.HIGH,
        "description": "条款包含单方解除权",
    },
    {
        "name": "unfair_penalty",
        "keywords": ["违约金不低于", "双倍赔偿", "三倍赔偿", "全额赔偿"],
        "risk_level": RiskLevel.MEDIUM,
        "description": "违约金条款可能不合理",
    },
    {
        "name": "auto_renewal",
        "keywords": ["自动续约", "自动续期", "默认续签"],
        "risk_level": RiskLevel.MEDIUM,
        "description": "包含自动续约条款",
    },
    {
        "name": "vague_terms",
        "keywords": ["酌情处理", "另行协商", "视情况而定", "合理范围"],
        "risk_level": RiskLevel.LOW,
        "description": "条款表述模糊，存在歧义风险",
    },
    {
        "name": "exclusive_jurisdiction",
        "keywords": ["仅由.*法院管辖", "排他性管辖"],
        "risk_level": RiskLevel.MEDIUM,
        "description": "管辖权条款可能对一方不利",
    },
]


class RiskIdentificationAgent:
    """风险识别Agent：规则预筛 + 大模型深度分析。"""

    def __init__(self):
        self.llm = None

    def _ensure_llm(self):
        if self.llm is None:
            self.llm = get_llm()

    def __call__(self, state: ContractReviewState) -> dict:
        """LangGraph节点函数：执行风险识别。"""
        logger.info("风险识别Agent开始处理 review_id=%s", state.review_id)

        if not state.clauses:
            return {
                "risk_findings": [],
                "overall_risk_level": RiskLevel.NONE,
                "risk_summary": "无条款可供分析",
            }

        try:
            rule_findings = self._rule_based_scan(state.clauses)

            self._ensure_llm()
            llm_findings = self._llm_analysis(state.clauses)

            merged = self._merge_findings(rule_findings, llm_findings)
            overall = self._calculate_overall_risk(merged)
            summary = self._generate_summary(merged, overall)

            logger.info(
                "风险识别完成：%d个发现，整体风险=%s",
                len(merged), overall.value,
            )

            return {
                "risk_findings": merged,
                "overall_risk_level": overall,
                "risk_summary": summary,
                "needs_human_review": overall == RiskLevel.HIGH,
            }

        except Exception as e:
            logger.error("风险识别Agent异常: %s", str(e))
            rule_findings = self._rule_based_scan(state.clauses)
            overall = self._calculate_overall_risk(rule_findings)
            return {
                "risk_findings": rule_findings,
                "overall_risk_level": overall,
                "risk_summary": f"大模型分析失败，仅展示规则引擎结果。错误: {str(e)}",
                "errors": state.errors + [f"风险识别LLM分析失败: {str(e)}"],
            }

    def _rule_based_scan(self, clauses: list[Clause]) -> list[RiskFinding]:
        """第一层：规则引擎快速扫描。"""
        findings = []
        for clause in clauses:
            for rule in RISK_RULES:
                for keyword in rule["keywords"]:
                    if keyword in clause.content:
                        findings.append(
                            RiskFinding(
                                clause_id=clause.id,
                                risk_level=rule["risk_level"],
                                risk_type=rule["name"],
                                description=f"{rule['description']}（命中关键词：{keyword}）",
                                rationale=f"规则引擎检测：条款「{clause.title}」中包含「{keyword}」",
                            )
                        )
                        break
        return findings

    def _llm_analysis(self, clauses: list[Clause]) -> list[RiskFinding]:
        """第三层：大模型深度语义分析。"""
        clauses_text = "\n\n".join(
            f"[条款ID: {c.id}] {c.title}\n{c.content}" for c in clauses
        )

        messages = [
            SystemMessage(content=SYSTEM_PROMPT),
            HumanMessage(content=f"请分析以下合同条款的风险：\n\n{clauses_text}"),
        ]
        response = self.llm.invoke(messages)
        content = response.content.strip()

        if content.startswith("```"):
            content = content.split("\n", 1)[1]
            if content.endswith("```"):
                content = content[:-3]

        result = json.loads(content)
        findings = []
        for item in result.get("findings", []):
            try:
                level = RiskLevel(item.get("risk_level", "none"))
            except ValueError:
                level = RiskLevel.LOW
            findings.append(
                RiskFinding(
                    clause_id=item.get("clause_id", ""),
                    risk_level=level,
                    risk_type=item.get("risk_type", "unknown"),
                    description=item.get("description", ""),
                    buyer_impact=item.get("buyer_impact", ""),
                    seller_impact=item.get("seller_impact", ""),
                    rationale=item.get("rationale", ""),
                )
            )
        return findings

    def _merge_findings(
        self, rule_findings: list[RiskFinding], llm_findings: list[RiskFinding]
    ) -> list[RiskFinding]:
        """合并规则引擎和大模型的结果，去重并取最高风险等级。"""
        seen: dict[tuple[str, str], RiskFinding] = {}

        for f in rule_findings:
            key = (f.clause_id, f.risk_type)
            seen[key] = f

        priority = {RiskLevel.HIGH: 3, RiskLevel.MEDIUM: 2, RiskLevel.LOW: 1, RiskLevel.NONE: 0}
        for f in llm_findings:
            key = (f.clause_id, f.risk_type)
            if key in seen:
                existing = seen[key]
                if priority[f.risk_level] > priority[existing.risk_level]:
                    seen[key] = f
            else:
                seen[key] = f

        return list(seen.values())

    def _calculate_overall_risk(self, findings: list[RiskFinding]) -> RiskLevel:
        if any(f.risk_level == RiskLevel.HIGH for f in findings):
            return RiskLevel.HIGH
        if any(f.risk_level == RiskLevel.MEDIUM for f in findings):
            return RiskLevel.MEDIUM
        if any(f.risk_level == RiskLevel.LOW for f in findings):
            return RiskLevel.LOW
        return RiskLevel.NONE

    def _generate_summary(self, findings: list[RiskFinding], overall: RiskLevel) -> str:
        high = sum(1 for f in findings if f.risk_level == RiskLevel.HIGH)
        medium = sum(1 for f in findings if f.risk_level == RiskLevel.MEDIUM)
        low = sum(1 for f in findings if f.risk_level == RiskLevel.LOW)
        return (
            f"整体风险等级：{overall.value}。"
            f"发现 {high} 个高风险、{medium} 个中风险、{low} 个低风险问题。"
        )
